### PR TITLE
Lazy Profiling + Debug "lazy-trace"

### PR DESCRIPTION
**Pull request by SkySkimmer**
also Ccnv Profiling

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by Janno** [comment]
This is cool! I've been trying to see if I can learn anything from the output. The output of `lazy-trace` seems to contain a lot of useful information but it is very very hard to parse without building a proper tool for it. The automatically inserted line breaks really work against my usual approach of throwing a bunch of `sed` and `rg` at it. It would be very helpful to have a json based output (or anything else that offers tools comparable to `jq`). 

The way I envision using this is to pose queries such as "how match steps happen directly inside this function", "how often do we reduce function A in function B", "is there a transitive call chain from function C to function D" etc. 

I have managed to at least get the entire stack on one line. I guess I could now use a bunch more grepping to answer these kinds of questions.
```bash
cat /tmp/trace-qed.out | rg -U --multiline-dotall '<infomsg>.*?Debug:.*?\[lazy-trace\].*?(.*?)</infomsg>' -r '$1' -o --json | sed -re 's,\\n,,g' | sed -re 's,  +, ,g' | rg '<infomsg>.*?Debug:.*?\[lazy-trace\].*?(.*?)</infomsg>' -r '$1' -o | sed -re 's,^ ,,'
```

It might also be nice to expose control over the counts and traces to clients. I have calls to kernel reduction that really should be grouped when it comes to statistics and traces. It would also be very easy to do some kind of processing on the traces directly in OCaml.

I've also run performance tests on BlueRock code. Somewhat expectedly, it impacts our generated files (one huge `Definition`) more than our proofs (lots of time in tactic code).

| Relative | Master   | MR       | Change   | Filename
|---------:|---------:|---------:|---------:|----------
|   +0.79% |  35975.3 |  36260.5 |   +285.1 | cpp2v-generated
|   +0.16% |  87992.2 |  88129.6 |   +137.5 | other
|   +0.34% | 123967.5 | 124390.1 |   +422.6 | total

Some more thoughts:

Mtac2 had tracing that looked quite similar to this approach. (@ppedrot helped me figure out how to time things correctly in the crazy call-by-name semantics of Mtac2.) The tracing was limited to just delta steps and Mtac2 primitives because almost everything else is handled directly by the kernel. Most importantly, it also recorded time stamps which I used it to generate flame graphs (see very old example below). Those were very helpful to figure out where Mtac2 spends its time. I wonder if it would be too much overhead to record timestamps in Coq. Unfortunately it is possible that getting the current time at a useful resolution adds noticeable overhead that could skew the results.

<details><summary>Example SVG flame graph of an Mtac2 program</summary>

heap_lang

</details>

On a wilder note, and maybe as an alternative to recording timestamps that won't skew the results: `gdb` has support for stopping programs regularly and poking around in their state. It can be used to take stack samples and it could also be used to sample the kernel trace added by this MR *if* the trace was exposed at a fixed or somehow discoverable address. Finding the trace is probably the hardest part with this idea. Writing a bit of `gdb` script code to traverse an OCaml list should be doable. Translating the names to strings might be more difficult but I am positive that some solution could be figured out.

The reason I am suggesting this is that if we can get the timestamps to agree with `perf`'s timestamps (using `CLOCK_MONOTONIC`, probably) we could finally get performance results for `Qed` that align the OCaml code with the Coq functions being reduced.

**Comment by SkySkimmer** [comment]
>I've also run performance tests on BlueRock code. 

I guess this is with the debug flags off ie only measures the cost when not using the feature?

**Comment by Janno** [comment]
> I guess this is with the debug flags off ie only measures the cost when not using the feature?

Yes, this is just using the current code of the MR. If you want I can tweak the code to make it perform the work of recording without actually printing anything (because that would probably wreck our CI logs) and rerun the measurements.

**Comment by SkySkimmer** [comment]
@coqbot bench

@Janno I pushed

**Comment by Janno** [comment]
| Relative | Master   | MR       | Change   | Filename
|---------:|---------:|---------:|---------:|----------
|   +5.71% |  35975.3 |  38027.8 |  +2052.4 | cpp2v-generated
|   +0.89% |  87991.9 |  88774.7 |   +782.8 | other
|   +2.29% | 123967.2 | 126802.4 |  +2835.2 | total

**Comment by SkySkimmer** [comment]
seems like that would do a lot of extra allocation

**Comment by Janno** [comment]
I tried collecting events in a list and it seems slower than the hashtable. Despite the fact that I fixed the problem with work happening in `print_recorded_steps` when nothing is supposed to be printed. 

Here's my change: https://github.com/coq/coq/commit/8447d30f5983612aa11aff2ab21194914caa4cfc. 

Results on our code base (relative to master without this MR, not relative to the previous version of the MR):

| Relative | Master   | MR       | Change   | Filename
|---------:|---------:|---------:|---------:|----------
|   +6.73% |  35975.4 |  38397.4 |  +2422.0 | cpp2v-generated
|   +1.70% |  87991.9 |  89488.8 |  +1496.9 | other
|   +3.16% | 123967.2 | 127886.2 |  +3919.0 | total

**Comment by SkySkimmer** [comment]
>`  let create : t = ref []`

that seems wrong, you're accumulating to a global ref instead of creating one for the given `clos_tab`

**Comment by SkySkimmer** [comment]
>I think the fact that this call is unconditional really messes with the latest commit. 

fixed

@coqbot bench

**Comment by SkySkimmer** [comment]
>It's still pretty bad, though, absolutely speaking.

The standard way to go faster when profiling would be to record less than 100% of the steps

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by Janno** [comment]
Is there any way we can guard the changes by some kind of compile time flag and merge it? I have rebased dozens of branches on top of this one for debugging purposes and I got a lot of use out of it. But it is a big hassle. With a compile time flag at least the rebasing would become unnecessary. Maybe for opam-installed rocq the flag could be enabled by an optional opam dependency on a dummy package that can be installed to trigger recompilation.

**Comment by SkySkimmer** [comment]
Do we need a compile time flag? it's been a while so I don't remember if there was much cost when the option is off

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
Not sure why the bot didn't post this (bench with f8bcb7b02ed692abe67306ec63724df157f333be)
~~~
┌────────────────────────────────────┬─────────────────────────┬─────────────────────────────────────┬─────────────────────────┐
│                                    │      user time [s]      │          CPU instructions           │  max resident mem [KB]  │
│                                    │                         │                                     │                         │
│            package_name            │   NEW      OLD    PDIFF │      NEW            OLD       PDIFF │   NEW      OLD    PDIFF │
├────────────────────────────────────┼─────────────────────────┼─────────────────────────────────────┼─────────────────────────┤
│                rocq-mathcomp-order │  101.18   101.40  -0.22 │  687418554750   686288682877   0.16 │  919480   920320  -0.09 │
│                 rocq-mathcomp-boot │   41.13    41.10   0.07 │  231973788808   231319173471   0.28 │  667580   667720  -0.02 │
│                       rocq-runtime │   76.51    76.38   0.17 │  558234324109   556916916025   0.24 │  544044   543592   0.08 │
│         coq-performance-tests-lite │  892.98   891.19   0.20 │ 6933861012409  6909816746364   0.35 │ 1527824  1586584  -3.70 │
│                          rocq-elpi │   18.05    18.01   0.22 │  119683685337   119628167180   0.05 │  492300   492364  -0.01 │
│              rocq-mathcomp-algebra │  378.85   377.58   0.34 │ 2575807931315  2565010421403   0.42 │ 1563500  1564404  -0.06 │
│ rocq-mathcomp-group-representation │  100.02    99.40   0.62 │  663596530872   657201097760   0.97 │ 1522784  1521656   0.07 │
│                rocq-mathcomp-field │  212.25   210.90   0.64 │ 1431638972508  1424483668258   0.50 │ 2219708  2226208  -0.29 │
│                        rocq-stdlib │  250.86   249.20   0.67 │ 1457543412623  1451009917605   0.45 │  753712   755768  -0.27 │
│                           coq-hott │  170.35   169.16   0.70 │ 1052445700080  1047467697252   0.48 │  457372   474600  -3.63 │
│             rocq-mathcomp-analysis │ 1408.64  1397.12   0.82 │ 9863175385137  9786111483543   0.79 │ 2196244  2194692   0.07 │
│         rocq-mathcomp-finite-group │   27.80    27.57   0.83 │  173379426342   170932004664   1.43 │  575884   572200   0.64 │
│                       rocq-bignums │   26.17    25.94   0.89 │  157856694863   156929309256   0.59 │  455212   455128   0.02 │
│                          rocq-core │    9.64     9.55   0.94 │   61494619483    61372699241   0.20 │  549844   549436   0.07 │
│                           coq-core │    2.81     2.78   1.08 │   19263154445    19257946600   0.03 │  114492   114200   0.26 │
│             rocq-mathcomp-solvable │  102.86   101.73   1.11 │  659441258499   652377807396   1.08 │ 1085572  1076632   0.83 │
│             coq-mathcomp-odd-order │  632.48   623.79   1.39 │ 4247583401697  4189383674891   1.39 │ 2731592  2715924   0.58 │
│              coq-engine-bench-lite │  130.79   128.19   2.03 │  949819026305   935211288819   1.56 │ 1152832  1145636   0.63 │
│            rocq-mathcomp-ssreflect │    1.24     1.21   2.48 │    7439805924     7447990790  -0.11 │  612540   605936   1.09 │
└────────────────────────────────────┴─────────────────────────┴─────────────────────────────────────┴─────────────────────────┘
~~~

**Comment by SkySkimmer** [comment]
added to https://github.com/rocq-prover/rocq/wiki/Rocq-Call-2026-09-22

**Comment by mattam82** [comment]
@Janno we discussed this at the Rocq call, and we would like to ask: is it possible to make it more efficient when it's not on (do we know if it's memory usage or the record_step calls maybe?) and what a compile-time option (a la ppx_optcomp) would do to the code?

**Comment by Janno** [comment]
There are some small things that can be done to slightly mitigate the slowdown without going to conditional compilation. For example, `record_step` does some work even when profiling is disabled. But fixing all those only reduces the overhead by 1/5 in my benchmarks.

I think the biggest slowdown comes from
*  the size increase in `fconstr` and `stack_member` and the resulting GC pressure
* and the extra function arguments everywhere.

I suspect we'll need conditional compilation to get rid of either of those problems. But even then I do not see an easy solution. I had a look at `ppx_optcomp` and `cppo` and did not like the trade-offs.

AFAICT `ppx_optcomp` can only change entire top-level blocks. We would have to copy a whole lot of code between the two versions. There is also possibly a problem with compatibility with OCaml 4.14 (only supported by v0.16 of the package and that comes with dependencies on jane street libraries). It also does not seem to support 32bit architectures but I am not 100% sure. On the plus side, it probably is easier to deal with in editors since merlin should know how to handle ppx and the annotations are standard syntax.

`cppo` can make finer-grained changes that avoid copying too much of the code. But I would absolutely hate to maintain the resulting files. I pushed a candidate here: https://github.com/Janno/coq/commit/548cab0290a965c928d6b9b1770cb7e3c81ad8ad#diff-6ba42b3fedb527be11adfc381b34a43ba95809e9f938b2cf34b96c4d71f6e0a7 so you can see for yourself. I am pretty sure no editor will be able to deal with these. And merlin does not know how to handle these files either.

I think that there might also be a way to get by with functors to get a profiling and a non-profiling version. This would be at the cost of making the code much more convoluted, some duplication due to signatures, and some actual duplication because we cannot abstract over extra fields in records.

Do any of these solutions sound appealing?

**Comment by SkySkimmer** [comment]
I'm not sure functors inline enough to make it efficient

**Comment by SkySkimmer** [comment]
I think conditional compilation of the functions to remove the extra argument is probably too unreadable (unless we write our own ppx maybe, if it's possible to have a ppx do something like `let f x y [%condarg:ctx] = g y [%condarg:ctx] x`)

It may be possible to use conditional compilation locally to have the small fconstr for disabled profiling, something like implementing
~~~ocaml
type ('t,'m,'c) fconstr0 (* = { mark:'m; term:'t } when no profile, = { mark:'m; term:'t; ctx:'c option } when profile *)
val mkf : 'c option -> 'm -> 't -> ('t,'m,'c) fconstr0 (* drop the 'c option argument when no profile *)
val fterm :  ('t,_,_) fconstr0 -> 't 
(* etc *)

val fctx : (_,_,'c) fconstr0 -> 'c option (* always None when no profile *)
~~~
and then `type fconstr = (fterm,mark,current_context) fconstr0 and fterm = ...` is unconditional.
as long as mkf fterm etc get inlined it should be as efficient as directly `type fconstr = { ... }` I think 

That would leave some fattening in stack_member and extra function arguments I guess
